internal/controller/controllers/crd_utils.go

```diff
@@ -55,48 +63,51 @@ func (u *CRDUtils) CreateAgentCR(ctx context.Context, log logrus.FieldLogger, ho
 	}
 
 	if err != nil && k8serrors.IsNotFound(err) {
-		labels := map[string]string{aiv1beta1.InfraEnvNameLabel: infraEnv.Name}
-		if infraEnv.Spec.AgentLabels != nil {
-			for k, v := range infraEnv.Spec.AgentLabels {
+		labels := map[string]string{aiv1beta1.InfraEnvNameLabel: infraEnvCR.Name}
+		if infraEnvCR.Spec.AgentLabels != nil {
+			for k, v := range infraEnvCR.Spec.AgentLabels {
 				labels[k] = v
 			}
 		}
 
 		host = &aiv1beta1.Agent{
 			Spec: aiv1beta1.AgentSpec{
-				ClusterDeploymentName: &aiv1beta1.ClusterReference{
-					Name:      clusterName,
-					Namespace: clusterNamespace,
-				},
 				Approved: false,
 			},
 			ObjectMeta: metav1.ObjectMeta{
 				Name:      hostId,
-				Namespace: infraEnv.Namespace,
+				Namespace: infraEnvCR.Namespace,
 				Labels:    labels,
 			},
 		}
 
+		if cluster != nil && cluster.KubeKeyNamespace != "" {
+			host.Spec.ClusterDeploymentName = &aiv1beta1.ClusterReference{
+				Name:      cluster.KubeKeyName,
+				Namespace: cluster.KubeKeyNamespace,
+			}
+		}
+
 		// Update 'kube_key_namespace' in Host
-		if err1 := u.hostApi.UpdateKubeKeyNS(ctx, hostId, infraEnv.Namespace); err1 != nil {
+		if err1 := u.hostApi.UpdateKubeKeyNS(ctx, hostId, infraEnvCR.Namespace); err1 != nil {
 			return errors.Wrapf(err1, "Failed to update 'kube_key_namespace' in host %s", hostId)
 		}
 
-		log.Infof("Creating Agent CR. Namespace: %s, Cluster: %s, HostID: %s", infraEnv.Namespace, clusterName, hostId)
+		log.Infof("Creating Agent CR. Namespace: %s, Cluster: %s, HostID: %s", infraEnvCR.Namespace, clusterName, hostId)
 		return u.client.Create(ctx, host)
 	}
 
 	if err == nil {
 		log.Infof("Agent CR %s already exists", hostId)
-		key := types.NamespacedName{Name: hostId, Namespace: infraEnv.Namespace}
+		key := types.NamespacedName{Name: hostId, Namespace: infraEnvCR.Namespace}
 		// fetch previous by KubeKey
 		h, err2 := u.hostApi.GetHostByKubeKey(key)
 		if err2 != nil && !errors.Is(err2, gorm.ErrRecordNotFound) {
 			return errors.Wrapf(err2, "Failed to GetHostByKubeKey Name: %s Namespace: %s", key.Name, key.Namespace)
 		}
 
 		if err2 == nil {
-			if *h.ClusterID == *clusterID {
+			if cluster != nil && h.ClusterID != nil && *h.ClusterID == *cluster.ID {
 				log.Infof("Agent CR %s already exists, same cluster %s", hostId, h.ClusterID)
```

Comment:
if the host belong to infra-env only (if it is in pool) then does not it means that we will deregister it at line 115?

Reply:
Good catch. Added a test and fix.